### PR TITLE
regpg: init at 1.11

### PR DESCRIPTION
**Pull request by 0xC45**
###### Description of changes

Add new package "regpg": https://dotat.at/prog/regpg/

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/security/regpg/default.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  version = "1.11";
+
+  src = fetchgit {
+    url = "https://github.com/fanf2/regpg.git";
+    rev = "refs/tags/regpg-${version}";
+    sha256 = "9492edfa89d044c37a72f48e5753c0054fdf7f1100c834c8eadf518ce93d0d1c";
+    leaveDotGit = true;
+  };
+
+  builder = builtins.toFile "builder.sh" ''
```
Comment (symphorien) [inline pkgs/tools/security/regpg/default.nix:20]: please use the default builder. It has all sorts of niceties that are easy to forget when making your own.
Add git and perl to the path with `nativeBuildInputs = [ git perl ]`
run patchShebang in postPatch and postBuild
etc

File: pkgs/tools/security/regpg/default.nix [hunk 2]
```diff
@@ -0,0 +1,56 @@
+{ lib
+, stdenv
+, fetchgit
+, git
+, perl
+, perlTextMarkdown
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:6]:
```suggestion
, TextMarkdown
```

File: pkgs/tools/security/regpg/default.nix (hunk 1 above)
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:20]: Replace this with various `pre/post`Phases

File: pkgs/tools/security/regpg/default.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+}:
+
+stdenv.mkDerivation rec {
+  pname = "regpg";
+  version = "1.11";
+
+  src = fetchgit {
+    url = "https://github.com/fanf2/regpg.git";
+    rev = "refs/tags/regpg-${version}";
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:15]: Use `fetchFromGitHub`

File: pkgs/tools/security/regpg/default.nix [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  meta = with lib; {
+    description = "GPG wrapper utility for storing secrets in VCS";
+    homepage    = "https://dotat.at/prog/regpg/";
+    license     = licenses.gpl3;
+    platforms   = platforms.all;
+    maintainers = with maintainers; [ "0xc45" ];
+  };
+
+  inherit git perl perlTextMarkdown;
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:55]: This is also a no-no.
```suggestion
```

File: pkgs/tools/security/regpg/default.nix [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    url = "https://github.com/fanf2/regpg.git";
+    rev = "refs/tags/regpg-${version}";
+    sha256 = "9492edfa89d044c37a72f48e5753c0054fdf7f1100c834c8eadf518ce93d0d1c";
+    leaveDotGit = true;
+  };
+
+  builder = builtins.toFile "builder.sh" ''
+    source $stdenv/setup
+
+    export PATH="$git/bin:$perl/bin:$PATH"
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:23]: Use `lib.makeBinPath`

File: pkgs/tools/security/regpg/default.nix [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    export PERL5LIB="$perlTextMarkdown/lib/perl5/site_perl"
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:24]: .. and `makePerlPath`

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -21627,6 +21629,8 @@ with pkgs;
 
   perlcritic = perlPackages.PerlCritic;
 
+  perlTextMarkdown = perlPackages.TextMarkdown;
+
```
Comment (ckiee) [inline pkgs/top-level/all-packages.nix:21633]: Now redundant

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -9977,6 +9977,8 @@ with pkgs;
 
   reftools = callPackage ../development/tools/reftools { };
 
+  regpg = callPackage ../tools/security/regpg { };
```
Comment (ckiee) [inline pkgs/top-level/all-packages.nix:9980]:
```suggestion
  regpg = callPackage ../tools/security/regpg {
    inherit (perlPackages) TextMarkdown;
  };
```

File: pkgs/tools/security/regpg/default.nix [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    # Switch to nix perl (for executable)
+    patchShebangs --build ./regpg
+
+    make prefix="" DESTDIR=$out install
+  '';
+
+  meta = with lib; {
+    description = "GPG wrapper utility for storing secrets in VCS";
+    homepage    = "https://dotat.at/prog/regpg/";
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:49]:
```suggestion
    homepage    = "https://dotat.at/prog/regpg";
```

File: pkgs/tools/security/regpg/default.nix [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    license     = licenses.gpl3;
+    platforms   = platforms.all;
+    maintainers = with maintainers; [ "0xc45" ];
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:52]: Add yourself to `maintainers/maintainer-list.nix`, and then:
```suggestion
    maintainers = with maintainers; [ 0xc45 ];
```

**Review by ckiee — CHANGES_REQUESTED** [review]
Also run `nixpkgs-fmt`.

File: pkgs/tools/security/regpg/default.nix (hunk 2 above)
Comment (pennae) [reply pkgs/tools/security/regpg/default.nix:6]: or prefer to build a custom perl env with `perl.withPackages`, which probably be even better.

File: pkgs/tools/security/regpg/default.nix (hunk 10 above)
Comment (pennae) [reply pkgs/tools/security/regpg/default.nix:52]: more like
```suggestion
    maintainers = with maintainers; [ _0xc45 ];
```
and needs an addition to `maintainers.nix`

File: pkgs/tools/security/regpg/default.nix (hunk 5 above)
Comment (pennae) [reply pkgs/tools/security/regpg/default.nix:23]: with the default builder this will be entirely redundant

File: maintainers/maintainer-list.nix [hunk 11]
```diff
@@ -14669,4 +14669,10 @@
     github = "npatsakula";
     githubId = 23001619;
   };
+  _0xC45 = {
+    email = "jason@0xc45.com";
+    name = "Jason Vigil";
+    github = "0xC45";
+    githubId = 56617252;
+  };
```
Comment (pennae) [inline maintainers/maintainer-list.nix:14677]: this should be added in a separate commit (`maintainers: add 0xC45`) and added to the top of the file in the properly sorted place

File: pkgs/tools/security/regpg/default.nix [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    rev = "regpg-${version}";
+    sha256 = "2ea99950804078190e1cc2a76d4740e3fdd5395a9043db3f3fe86bf2477d3a7d";
+  };
+
+  nativeBuildInputs = [ git perlEnv ];
+
+  postPatch = ''
+    patchShebangs ./
+    substituteInPlace util/insert-here.pl \
+      --replace 'qx(git describe)' '"${version}"'
```
Comment (pennae) [inline pkgs/tools/security/regpg/default.nix:28]: `git describe` outputs the tag name, so this should be `rev` as used in the fetcher rather than `version`

File: pkgs/tools/security/regpg/default.nix [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+, stdenv
+, fetchFromGitHub
+, git
+, perl
+}:
+
+let
+  perlEnv = perl.withPackages (p: [ p.TextMarkdown ]);
+
+in
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:11]:
```suggestion
  perlEnv = perl.withPackages (p: [ p.TextMarkdown ]);
in

```

File: pkgs/tools/security/regpg/default.nix [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    owner = "fanf2";
+    repo = "regpg";
+    rev = "regpg-${version}";
+    sha256 = "2ea99950804078190e1cc2a76d4740e3fdd5395a9043db3f3fe86bf2477d3a7d";
+  };
+
+  nativeBuildInputs = [ git perlEnv ];
+
+  postPatch = ''
+    patchShebangs ./
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:26]: This is slow over the entire source tree, try to narrow it down more.
(Also, no trailing `/` is needed)

File: pkgs/tools/security/regpg/default.nix [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    patchShebangs --build ./regpg
+  '';
+
+  makeFlags = [ "prefix=$(out)" ];
+
+  meta = with lib; {
+    description = "GPG wrapper utility for storing secrets in VCS";
+    homepage = "https://dotat.at/prog/regpg";
+    license = licenses.gpl3;
+    platforms = platforms.all;
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:43]:
```suggestion
    inherit (perl.meta) platforms;
```

File: pkgs/tools/security/regpg/default.nix [hunk 16]
```diff
@@ -0,0 +1,46 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, git
+, perl
+}:
+
+let
+  perlEnv = perl.withPackages (p: [ p.TextMarkdown ]);
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:9]:
```suggestion
  perlEnv = perl.withPackages (p: with p; [ TextMarkdown ]);
```

File: maintainers/maintainer-list.nix [hunk 17]
```diff
@@ -78,6 +78,12 @@
       fingerprint = "2536 9E86 1AA5 9EB7 4C47  B138 6510 870A 77F4 9A99";
     }];
   };
+  _0xC45 = {
+    email = "jason@0xc45.com";
+    name = "Jason Vigil";
+    github = "0xC45";
+    githubId = 56617252;
+  };
```
Comment (ckiee) [inline maintainers/maintainer-list.nix:86]: Perhaps add your `matrix`? Would be nice to have since it's `0xc45` instead of `0xc45` there.

**Review by ckiee — CHANGES_REQUESTED** [review]
Just nit-picks now

File: pkgs/tools/security/regpg/default.nix (hunk 15 above)
Comment (ckiee) [reply pkgs/tools/security/regpg/default.nix:43]: Seems perl is `platforms.all` too, so I guess this is irrelevant.

File: pkgs/tools/security/regpg/default.nix [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+  version = "1.11";
+
+  src = fetchFromGitHub {
+    owner = "fanf2";
+    repo = "regpg";
+    rev = "regpg-${version}";
+    sha256 = "2ea99950804078190e1cc2a76d4740e3fdd5395a9043db3f3fe86bf2477d3a7d";
+  };
+
+  nativeBuildInputs = [ git perlEnv ];
```
Comment (pennae) [inline pkgs/tools/security/regpg/default.nix:23]: is git still needed? (haven't checked the source)

**Review by pennae — COMMENTED** [review]
the main script of this package probably needs a wrapper that inserts `gpg` into its path. without such a wrapper you'd need to have gpg in your system path for this to work (which is likely to be the case, but we shouldn't depend on it)

**Comment by 0xC45** [comment]
> the main script of this package probably needs a wrapper that inserts `gpg` into its path. without such a wrapper you'd need to have gpg in your system path for this to work (which is likely to be the case, but we shouldn't depend on it)

Ah, good point. Done.

**Comment by 0xC45** [comment]
Actually, just noticed darwin builds were broken. Maybe this will fix it.

File: pkgs/tools/security/regpg/default.nix [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    repo = "regpg";
+    rev = "regpg-${version}";
+    sha256 = "2ea99950804078190e1cc2a76d4740e3fdd5395a9043db3f3fe86bf2477d3a7d";
+  };
+
+  nativeBuildInputs = [ makeWrapper perlEnv ];
+
+  postPatch = ''
+    patchShebangs ./util/insert-here.pl
+    patchShebangs ./util/markdown.pl
```
Comment (ckiee) [inline pkgs/tools/security/regpg/default.nix:27]:
```suggestion
    patchShebangs ./util/insert-here.pl ./util/markdown.pl
```

**Review by ckiee — APPROVED** [review]
Very nice for a first contribution (:

**Comment by robockie** [comment]
Result of `nixpkgs-review pr 178526` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>regpg</li>
  </ul>
</details>
